src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+    // TASK_BUFN_THROW and TASK_SPLIT_THROW. They should never happen unless this is being called
+    // from within an alloc.
+    while (!done) {
+      auto thread = threads.find(thread_id);
+      if (thread != threads.end()) {
+        switch (thread->second.state) {
+        case TASK_BLOCKED:
+        // fall through
+        case TASK_BUFN:
+        // fall through
+        case SHUFFLE_BLOCKED:
+          log_status("WAITING", thread_id, thread->second.task_id, thread->second.state);
+          task_wake_condition.wait(lock);
+          task_has_woken_condition.notify_all();
+          break;
+        case SHUFFLE_THROW:
+          transition(thread->second, thread_state::SHUFFLE_RUNNING);
+          throw_java_exception(JAVA_OOM_CLASS, "could not allocate enough for shuffle");
+          break;
+        case TASK_BUFN_THROW:
+          transition(thread->second, thread_state::TASK_BUFN_WAIT);
+          throw_retry_oom("rollback and retry operation", thread->second, lock);
+          break;
+        case TASK_BUFN_WAIT:
+          transition(thread->second, thread_state::TASK_BUFN);
+          log_status("WAITING", thread_id, thread->second.task_id, thread->second.state);
+          task_wake_condition.wait(lock);
+          task_has_woken_condition.notify_all();
+          break;
+        case TASK_SPLIT_THROW:
+          transition(thread->second, thread_state::TASK_RUNNING);
+          throw_split_n_retry_oom("rollback, split input, and retry operation", thread->second, lock);
+          break;
+        case TASK_REMOVE_THROW:
+        // fall through
+        case SHUFFLE_REMOVE_THROW:
+          log_transition(thread_id, thread->second.task_id, thread->second.state, thread_state::UNKNOWN);
+          threads.erase(thread);
+          task_has_woken_condition.notify_all();
+          throw std::runtime_error("thread removed while blocked");
+        default:
+          if (!first_time) {
+            log_status("DONE WAITING", thread_id, thread->second.task_id, thread->second.state);
+          }
+          done = true;
+        }
+      }
+      first_time = false;
+    }
+  }
+
+  /**
+   * Wake up threads after a task finished. The task finishing successfully means
+   * that progress was made. So we want to restart some tasks to see if they can
+   * make progress. Right now the idea is to wake up all blocked threads first
+   * and if there are no blocked threads, then we wake up all BUFN threads.
+   * Hopefully the frees have already woken up all the blocked threads anyways.
+   */
+  void wake_up_threads_after_task_finishes(std::unique_lock<std::mutex> & lock) {
+    bool are_any_tasks_just_blocked = false;
+    for (auto thread = threads.begin(); thread != threads.end(); thread++) {
+      switch(thread->second.state) {
+        case TASK_BLOCKED:
+          transition(thread->second, thread_state::TASK_RUNNING);
+          task_wake_condition.notify_all();
```

Comment:
It's a bit odd that we're notifying multiple times on the same condition variable, once per task in a blocked state.  Given the way the condition variable is being used and that I think we're always doing a state transition on the task before waking it up, seems like we could avoid a lot of unnecessary wakeups by having a condition variable per task rather than a global one.